src/Algebra/Graph/AdjacencyIntMap/Algorithm.hs

```diff
@@ -31,6 +35,63 @@ import qualified Data.Graph.Typed   as Typed
 import qualified Data.IntMap.Strict as IntMap
 import qualified Data.IntSet        as IntSet
 
+-- | Compute the forest of a graph's vertices in breadth first order.
+-- Complexity: /O(n+m*log n)/ time and /O(n+m)/ space.
+--
+-- @
+-- bfsForest 'empty'                         == []
+-- 'forest' (bfsForest $ 'edge' 1 1)           == 'vertex' 1
+-- 'forest' (bfsForest $ 'edge' 1 2)           == 'edge' 1 2
+-- 'forest' (bfsForest $ 'edge' 2 1)           == 'vertices' [1,2]
+-- 'isSubgraphOf' ('forest' $ bfsForest x) x   == True
+-- bfsForest . 'forest' . bfsForest          == bfsForest
+-- 'forest' (bfsForest ('circuit' [1..5] + 'circuit' [5,4..1])) == 'path' [1,2,3] + 'path' [1,5,4]
+-- @
+bfsForest :: AdjacencyIntMap -> Forest Int
+bfsForest g = bfsForestFrom' (vertexList g) g
+
+-- | Like 'bfsForest', but the traversal is seeded by a list of
+-- vertices. Vertices not in the graph are ignored.
+--
+-- Let /L/ be the number of seed vertices. Complexity: /O(n+(L+m)*logn)/
+-- time and /O(n+m)/ space.
+--
+-- @
+-- 'forest' (bfsForestFrom [1,2] $ 'edge' 1 2) == 'vertices' [1,2]
+-- 'forest' (bfsForestFrom [2]   $ 'edge' 1 2) == 'vertex' 2
+-- 'forest' (bfsForestFrom [3]   $ 'edge' 1 2) == empty
+-- 'forest' (bfsForestFrom [3] ('circuit' [1..5] + 'circuit' [5,4..1])) == 'path' [3,2,1] + 'path' [3,4,5]
+-- @
+bfsForestFrom :: [Int] -> AdjacencyIntMap -> Forest Int
+bfsForestFrom vs g = bfsForestFrom' [ v | v <- vs, hasVertex v g] g
+
+bfsForestFrom' :: [Int] -> AdjacencyIntMap -> Forest Int
+bfsForestFrom' vs g = evalState (bff vs) IntSet.empty where
+  bff [] = return []
+  bff (v:vs) = discovered v >>= \case
+    False -> bff vs
+    True -> (:) <$> unfoldTreeM_BF walk v <*> bff vs
+  walk v = (v,) <$> adjacentM v
+  adjacentM v = filterM discovered $ IntSet.toList (postIntSet v g)
+  discovered v = do new <- gets (not . IntSet.member v)
+                    when new $ modify' (IntSet.insert v)
+                    return new
+
+-- | Like 'bfsForestFrom' with the resulting forest converted to a
+--   level structure.  Flattening the result via @'concat' . 'bfs' vs@
+--   gives an enumeration of vertices reachable from @vs@ in breadth
+--   first order.
+--
+--   Let /L/ be the number of seed vertices. Complexity:
+--   /O(n+(L+m)*log n)/ time and /O(n+m)/ space.
+-- 
+-- @
+-- bfs [3] ('circuit' [1..5] + 'circuit' [5,4..1])          == [[3],[2,4],[1,5]]
+-- 'concat' (bfs [3] $ 'circuit' [1..5] + 'circuit' [5,4..1]) == [3,2,4,1,5]
```

Comment:
Is this always the case that vertices in a single level appear in sorted order? Probably not. In either case, I think it's useful to document this.

Reply:
Yeah, for 2 and 4, one of the levels is not in sorted order. I changed the test case and examples to start with 2 to illustrate this. I also added a remark in the documentation.